### ISSUE TITLE
House of the Dead III (D3D8) - Lisa's hair detaches from her head

### ISSUE THREAD

**Issue by SeongGino** [doitsujin/dxvk#4130]
### Software information
The House of the Dead III (PC)

### System information
- GPU: NVIDIA Geforce RTX 3060 Ti
- Driver: NVIDIA 545.29.02 (old), NVIDIA 555.58.02 (current)
- Wine version: Wine-GE-Proton8-23 (old), Wine-GE-Proton8-26 (current)
- DXVK version: D8VK 1.0 (old), DXVK master 4b0c9c6 (current)

### Apitrace file(s)
[hod3pc.5.trace.tar.xz](https://drive.google.com/file/d/1gFtPT4Ekd98H-oc_Mr6VgWi9T0LFYLNh/view?usp=sharing)

The first stage that the character appears, her hair detaches from her head - almost like it's frozen where it first appears in the shot.
Examples for visual aid, both from Stage 2:
![2023_11-22 195839](https://github.com/AlpyneDreams/d8vk/assets/7321839/c59b985d-d416-4fdf-a9f9-693142e3efe6)
![2023_11-22 195852](https://github.com/AlpyneDreams/d8vk/assets/7321839/a2d20a27-4ad3-495f-a6a9-3d56f4044a3f)

#### Edit: And though less noticeable, I believe some of the decals of the moss type enemies from Stage 1/Prologue slightly stick out unusually and persist beyond their deaths, similar to Lisa's hair.

NOTE: Though the apitrace is attached with WineD3D as instructed, some facial models are broken under it and I'm not sure if the resulting trace is reproduced perfectly:
(subtitles glitch here happens on D8VK too)
![2023_11-22 202805-THE HOUSE OF THE DEAD 3](https://github.com/AlpyneDreams/d8vk/assets/7321839/ef21c3f8-5abe-4928-9a73-ae6fa03b2b65)
![2023_11-22 202841-THE HOUSE OF THE DEAD 3](https://github.com/AlpyneDreams/d8vk/assets/7321839/b3436ab2-d79e-496e-8936-c827c371f2c3)
These issues aren't reproduced in dgvoodoo2-through-DXVK.

#### Note: this issue is re-posted from d8vk#193 (at the advice of @WinterSnowfall), but as can be seen from my last post on that thread at midnight today (1:38 AM @ July 9th, 2024), the issue is still present in the listed DXVK master version post-D8VK merge.
![2024_07-09 014406](https://github.com/AlpyneDreams/d8vk/assets/7321839/975fdbca-06ac-4d63-af7f-af04f256f114)

**Comment by WinterSnowfall** [comment]
To add some context, the offending section is somewhat into the game and unfortunately the game demo stops right before that part.

@Blisto91 has also determined that the "detached" hair is baked into the trace, as it can be observed even on Windows native while playing it back.

Since it appears to work fine on AMD post-upstreaming ([see here](https://github.com/AlpyneDreams/d8vk/issues/193#issuecomment-2214163846)), we would need to investigate with a copy of the game if this is perhaps a precision issue on Nvidia (I'm thinking floatEmulation and the like), or an outright bug.

**Comment by mrdeathjr28** [comment]
in my case with 2.4 works ok seems a nvidia problem ?

![thod3c](https://github.com/doitsujin/dxvk/assets/6601480/13bdcd50-f67c-4ac2-835d-03ab86c06a35)

![thod3d](https://github.com/doitsujin/dxvk/assets/6601480/2bb1d615-0398-44c3-a4c8-b1f2f7415a27)

![thod3e](https://github.com/doitsujin/dxvk/assets/6601480/3b37eb7b-8953-4f25-91eb-676e92cf4c64)

![thod3f](https://github.com/doitsujin/dxvk/assets/6601480/722033b9-0464-48d4-bebe-03456c48e8cd)

**Comment by WinterSnowfall** [comment]
@SeongGino If you have some time could you try a play-through of the problematic area with `d3d9.floatEmulation = Strict`? If you drop that line in a `dxvk.conf` file placed near the game's executable it's usually enough. You can verify in the logs that the option is properly parsed and applied; please make sure that is the case.

**Comment by SeongGino** [comment]
> @SeongGino If you have some time could you try a play-through of the problematic area with `d3d9.floatEmulation = Strict`? If you drop that line in a `dxvk.conf` file placed near the game's executable it's usually enough. You can verify in the logs that the option is properly parsed and applied; please make sure that is the case.

Noting the d3d9 in that setting, does that apply to D3D8 too?

I'll investigate before the end of the night. Sounds like a very obscure bug to be vendor specific.

**Comment by WinterSnowfall** [comment]
> Noting the d3d9 in that setting, does that apply to D3D8 too?

Yes, it does. You can use any of the d3d9 config options with d3d8 as well.
 
> Sounds like a very obscure bug to be vendor specific.

It wouldn't be the first time we've had one like this, if this is what it turns out to be.

**Comment by SeongGino** [comment]
~~Sorry about that, family related emergency.~~

Unfortunately, no change. Hair still brokey. :(
![2024_07-11 022926](https://github.com/doitsujin/dxvk/assets/7321839/7e28a102-1fef-4a3d-be3f-e151a667a293)
![2024_07-11 022550](https://github.com/doitsujin/dxvk/assets/7321839/a0dcab2f-3049-4681-a876-a23cddcfe5d1)

**Comment by WinterSnowfall** [comment]
> Unfortunately, no change. Hair still brokey. :(

Alright, thank you for ruling that out at least.

**Comment by dungeon007** [comment]
5+ years ago i was poking about this on  NINE + d3d8to9. Non-nvidia path works, but is like 3-4 times slower, so making nvidia path working is desirable.
 
![NINE](https://github.com/user-attachments/assets/d897ebb2-10d4-4e06-b3b5-69258cdd42b4)

 AFAIR i ended up hacking NINE and maded it work properly by removing instances of R5G6B5 format support. I guess app expect shorter display list or something in particular order there, otherwise gets corrupted and messes up.
 Dunno how everything is handled in DXVK, but my suggesting would be in that way... do something with 16-bit depth R5G6B5 format, so that app dont see it... something like that might help this nvidia path.

**Comment by WinterSnowfall** [comment]
Thank you for the insight. If that is indeed the case, I'd be happy to just use `d3d9.customVendorId = 1002` with the game, assuming that fixes the problem. Keep in mind this is a console port and most likely intended to run around 30 fps anyway.

**Comment by dungeon007** [comment]
Well, both ways, this text corruption bug still stays. 

![285079130-ef21c3f8-5abe-4928-9a73-ae6fa03b2b65](https://github.com/user-attachments/assets/7aadb2c3-b0eb-47c8-afaf-8d03a37accd1)

I dont remmeber this subtitles glitch... that must be dxvk specific.

**Comment by WinterSnowfall** [comment]
> I dont remember this subtitles glitch... that must be dxvk specific. But yeah in general, if app is messing something with a formats... then all of these glitches could be just false positive.

Right above, you can see a screenshot from @mrdeathjr28 where those subtitles appear to be fine? We'll try to procure the game and see if we can reproduce any of these things.

**Comment by SeongGino** [comment]
> Keep in mind this is a console port and most likely intended to run around 30 fps anyway.

The Xbox port of House of the Dead III runs at 60 (which makes sense; HOTD3 in arcades is Chihiro hardware, i.e. a ram doubled Xbox anyways)

**Comment by dungeon007** [comment]
> Right above, you can see a screenshot from @mrdeathjr28 where those subtitles appear to be fine? We'll try to procure the game and see if we can reproduce any of these things.
That text corruption seems happens only there and last for a second or two....so screenshot could misslead you if you take it a bit before or afterwards.

**Comment by SeongGino** [comment]
> > Right above, you can see a screenshot from @mrdeathjr28 where those subtitles appear to be fine? We'll try to procure the game and see if we can reproduce any of these things.
> 
> That text glitch seems happens only there and last for a second or two....so screenshot could misslead you if you take it a bit before or afterwards. And i didnt checked else languages, so maybe glitch is english subtitles specific

For what it's worth, neither the glitch subject of this issue nor that subtitles glitch happens when playing the game with dgVoodoo2 and rendering through DXVK's DX9 path.

**Comment by dungeon007** [comment]
Yeah, but dgVoodoo2 uses DX11 or DX12... that is entirely different path.

**Comment by mrdeathjr28** [comment]
> > For what it's worth, neither the glitch subject of this issue nor that subtitles glitch happens when playing the game with dgVoodoo2 and rendering through DXVK's DX9 path.
> 
> DXVK's d3d8 is mapped to d3d9, much like d3d8to9 wrapper. As mapped to different path, then they have different bugs potentional. Here bug is either in d3d8 and/or d3d9.

d3d8 is not exactlyd3d8to9 wrapper because have many things proper to dx8 compared d3d8to9 wrapper

however errors still present on 3d8to9 and dxvk

back to error in my case works ok, i finish a gameplay video around 3 hours (2 rescues) and have a glitch but different than posted @SeongGino and appear when you have finish game

![thod3ga](https://github.com/user-attachments/assets/23693de1-4086-4840-8861-1e2033c4d5dd)

![thod3gb](https://github.com/user-attachments/assets/97d2dba9-d408-40cb-aa30-d81bc90e72f4)

**Comment by dungeon007** [comment]
There are just so much "bugs" in this, it is not impossible that drivers became borked for this in a way too... somewhere back in history.

**Comment by K0bin** [comment]
> DXVK's D3D9 still need buffers optimisations, for SWVPs too, that still run TAD SLOW

SWVP should run fine unless a game does exceptionally weird stuff. It's implemented the same as HWVP except that SWVP supports more vertex constants. To avoid copying 1024 Vec4s for every draw call, we only copy up to the highest constant that gets modified by the application on the CPU and rely on robustness2 to return 0 for out of bounds reads. This approach has generally worked and SWVP stopped being a problem once we landed that, so I'd be surprised if its still slow in this game.

As for the bug, it sounds like it comes down to undocumented locking behavior. Maybe writing data after calling `Unlock()` or something like that.

**Comment by dungeon007** [comment]
@K0bin OK, i spot these optimisations since 2.3.1... as i see for Warrior Kings SWVPs this functioning, but yeah that is only partially. Since yeah i already know more cases of this turning bad, posted you a trace

**Comment by K0bin** [comment]
I don't know what you spot in 2.3.1 but the optimization I was talking about landed years ago.

https://github.com/doitsujin/dxvk/pull/2282

**Comment by dungeon007** [comment]
It is this i think ["Reworked how dynamic system memory buffers are copied to the GPU in D3D9 games."](https://github.com/doitsujin/dxvk/pull/3765)

**Comment by WinterSnowfall** [comment]
> About **d3d9.customVendorId** not functiong for d3d8 apps, that is third bug...

I've just checked with the d3d8-triangle test app, and it is being applied correctly, as determined by the d3d9 customVendorID logic. So are the rest of the fields (customDeviceId, customDeviceDesc). If you spot any particular points where a custom vendor ID isn't retrieved correctly, please open a bug with steps on how we can reproduce it.

Note that the dxvk UI will not report these frontend values, since it takes them from Vulkan directly.

**Comment by K0bin** [comment]
> It is this that triggered some SWVP perf i think https://github.com/doitsujin/dxvk/pull/3765

What does "triggered" mean here? Improved or worsened?

> Well, latest shader d3d version have 4 fog components, not 1.

No, it does not. The latest shader model which has fixed function fog is SM2_x and as you can see here:

https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx9-graphics-reference-asm-vs-registers-vs-2-x

oFog only has 1 component.

**Comment by dungeon007** [comment]
>What does "triggered" mean here? Improved or worsened?
Triggered means - even if didnt meant directly to improve it,  it get improved... partially 
Partially, means - not in every SWVP app. 
>oFog only has 1 component.
In NINE we do it 4 anyway, do not hurt. Maybe because newer shader compiler more likes it that way 🤣
